apps/studio/src/server/modules/auth/email/email.router.ts

```diff
@@ -45,37 +46,83 @@
 
       // May have one of them fail,
       // so users may get an email but not have the token saved, but that should be fine.
-      await Promise.all([
-        ctx.prisma.verificationToken.upsert({
-          where: {
-            identifier: email,
-          },
-          update: {
-            token: hashedToken,
-            expires,
-            attempts: 0,
-          },
-          create: {
-            identifier: email,
-            token: hashedToken,
-            expires,
-          },
-        }),
-        sendMail({
+      const { savedPrefix, newlyCreated } = await db
+        .transaction()
+        .setIsolationLevel("serializable")
+        .execute(async (tx) => {
+          const existing = await tx
+            .selectFrom("VerificationToken")
+            .forUpdate()
+            .where("identifier", "=", email)
+            .select(["expires", "prefix", "token"])
+            .executeTakeFirst()
+
+          // NOTE: if we don't have a stored entry for the token, we will create a new token
+          if (!existing) {
+            await tx
+              .insertInto("VerificationToken")
+              .values({
+                identifier: email,
+                token: hashedToken,
+                prefix: otpPrefix,
+                expires,
+              })
+              .execute()
+
+            return {
+              savedPrefix: otpPrefix,
+              newlyCreated: true,
+            }
+          } else if (existing.expires < new Date(Date.now())) {
```

Comment:
Should we have a buffer for this? Should be at least valid for 30 seconds? Else by the time the user checks and inputs it's expired

Reply:
removed this